polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -0,0 +1,1047 @@
+	/// ever)
+	pub(crate) fn hypothetical_membership(
+		&self,
+		candidate_hash: CandidateHash,
+		candidate: HypotheticalCandidate,
+		candidate_storage: &CandidateStorage,
+	) -> bool {
+		// If we've already used this candidate in the chain
+		if self.candidates.contains(&candidate_hash) {
+			return true
+		}
+
+		if !self.can_add_candidate_as_potential(
+			candidate_storage,
+			&candidate.relay_parent(),
+			candidate.parent_head_data_hash(),
+			candidate.output_head_data_hash(),
+		) {
+			return false
+		}
+
+		let Some(candidate_relay_parent) = self.scope.ancestor_by_hash(&candidate.relay_parent())
+		else {
+			// can_add_candidate_as_potential already checked for this, but just to be safe.
+			return false
+		};
+
+		let identity_modifications = ConstraintModifications::identity();
+		let cumulative_modifications = if let Some(last_candidate) = self.chain.last() {
+			&last_candidate.cumulative_modifications
+		} else {
+			&identity_modifications
+		};
+
+		let child_constraints =
+			match self.scope.base_constraints.apply_modifications(&cumulative_modifications) {
+				Err(e) => {
+					gum::debug!(
+						target: LOG_TARGET,
+						new_parent_head = ?cumulative_modifications.required_parent,
+						?candidate_hash,
+						err = ?e,
+						"Failed to apply modifications",
+					);
+
+					return false
+				},
+				Ok(c) => c,
+			};
+
+		let parent_head_hash = candidate.parent_head_data_hash();
+		if parent_head_hash == child_constraints.required_parent.hash() {
+			// We do additional checks for complete candidates.
+			if let HypotheticalCandidate::Complete { ref receipt, ref persisted_validation_data } =
+				candidate
+			{
+				let prospective_candidate = ProspectiveCandidate {
+					commitments: Cow::Borrowed(&receipt.commitments),
+					collator: receipt.descriptor().collator.clone(),
+					collator_signature: receipt.descriptor().signature.clone(),
+					persisted_validation_data: persisted_validation_data.as_ref().clone(),
+					pov_hash: receipt.descriptor().pov_hash,
+					validation_code_hash: receipt.descriptor().validation_code_hash,
+				};
+
+				if Fragment::new(
+					candidate_relay_parent.clone(),
+					child_constraints,
+					prospective_candidate,
+				)
+				.is_err()
+				{
+					gum::debug!(
+						target: LOG_TARGET,
+						"Fragment::new() returned error",
+					);
+					return false
+				}
+			}
+
+			// If we got this far, it can be added to the chain right now.
+			true
+		} else {
+			// Otherwise it is or can be an unconnected candidate.
+			true
+		}
+	}
+
+	/// Select `count` candidates after the given `ancestors` which pass
+	/// the predicate and have not already been backed on chain.
+	///
+	/// The intention of the `ancestors` is to allow queries on the basis of
+	/// one or more candidates which were previously pending availability becoming
+	/// available or candidates timing out.
+	pub(crate) fn find_backable_chain(
+		&self,
+		ancestors: Ancestors,
+		count: u32,
```

Comment:
Size of the chain we want 
```suggestion
		size: u32,
```

Reply:
or the candidate count :P

Reply:
Yeah, but wrt fn name size sounds proper